### DISCUSSION TITLE
[Enhancement] F# Support from Day 1

### DISCUSSION THREAD

**Discussion by staypufd** [dotnet/maui#115] [General]
## Summary
F# support from day 1 is critical. F# consistently gets left off of supported tooling with MS UI frameworks, this leaves it unsupported and it never recovers. Teams have made excuses for not supporting it and never seem to find the time or resources and thus F# support never happens. We’ve asked and pleaded for F# to truly be supported like the first class language it is held up to be. Please this time make it truly one from day 1 so that it isn’t behind from the very start and then dies as a option for so many who would use it if it was truly supported 100%. Please.

## Intended Use Case
Many of us functional developers use F# and want to use it for our GUI apps as well.

**Comment by nbevans** [comment]
Project system and templates is a bare minimum.

**Comment by Happypig375** [comment]
For inclusivity, add Visual Basic to the list as well.

**Reply by knocte** [reply]
VB has its own github issue: https://github.com/dotnet/maui/issues/266 , let's stick with F# on this one

**Comment by mrwcjoughin** [comment]
@Happypig375 vb.net is basically dead. I doubt it will even be supported in .net 7

**Comment by Padanian** [comment]
No doubt vb needs to be here. It's the 5th or 6th most used dev lang in the world, for both front ends and back ends, for CLI and GUI. 
F# is not necessary.

**Comment by mrwcjoughin** [comment]
@Padanian I don't believe that's accurate. While Microsoft is supporting it in .net 5, there is no commitment for anything after that, and they are flat out saying they are not adding new features to it at all:
https://visualstudiomagazine.com/articles/2020/03/12/vb-in-net-5.aspx

Bottom line: move to C# if you want to stay on .net - there really is no reason not to and so many reasons to move to it.

**Comment by Padanian** [comment]
I know what is their position _at this time_ . It's most probably going to change. 
And I'm not going to the curly bracket. Now back to the regular program.

**Comment by nbevans** [comment]
> No doubt vb needs to be here. It's the 5th or 6th most used dev lang in the world, for both front ends and back ends, for CLI and GUI.
> F# is not necessary.

Please don't say that. We have a 600k line codebase written in F# built on top Xamarin.Forms. If they remove F# support - we are toast. We would need to rewrite everything within the 12 months period they have allocated to continue supporting Xamarin.Forms.

For reference, Xamarin historically has somewhat encouraged use of F# on their platform and frameworks - Miguel de Icaza himself was pushing F# on the stack between at least 2014 and 2016 - they never did this so prominently for VB.NET. Then in 2018 the Xamarin team took on Don Syme (creator of F#) and the Fabulous MVU framework was produced - amongst other things internally. So F# *does* have a special place in Xamarin history and should continue to do so if for nothing else but maintaining backward compatibility.

In my view, all languages that Xamarin Forms supported should continue to be supported by MAUI with at least equitable functionality. This means project system and templates as a bare minimum.

**Comment by ncarandini** [comment]
> And I'm not going to the curly bracket.

Ok, that say it all. Good luck!

**Comment by brase** [comment]
@Padanian 
> And I'm not going to the curly bracket.

No problem. Try F# ;)

**Reply by saint4eva** [reply]
He is afraid f# is not going to be supported also. And he believes that vb.net is more popular than f# - maybe he is right. 🤷🏿

**Comment by 7sharp9** [comment]
It's not hard adding a template unless there is a lot of magic tooling occurring within.

**Comment by Krzysztof-Cieslak** [comment]
> No doubt vb needs to be here. [...]. F# is not necessary.

Don't have VB and F# communities bigger problems than shouting at each other?

**Comment by nbevans** [comment]
I support VB - they should experience no loss of functionality either in this Xamarin Forms to MAUI transition.

I don't think this should be a fight or choice between whether VB or F# is to be the lucky 2nd language that gets supported by MAUI. Let's just support all 3 to maximise the platform's reach and minimise arguments, pretty please?

**Comment by Happypig375** [comment]
This issue's like count has surpassed #34. Feels good man

**Comment by GiampaoloGabba** [comment]
I'm not (yet) an F# developer, i'm studying it a bit and is pure awesomeness :)
I'm sure that .net MAUI will bring some love to F#.

Regarding VB.NET i doubt it will get more than what it gets now: basically nothing. 
Some community effort will be ok, but i dont think (and hope) no one in the maui team will loose even a second to bring some love to VB.NET... i mean... come on...

**Comment by Padanian** [comment]
[By TIOBE index](https://www.tiobe.com/tiobe-index/), VB is the 6th largest community of developers at around 4%.
F# is the 46th with 0.2%. By _any standard_, it's a negligible minority.

**Comment by charlesroddie** [comment]
> Teams have made excuses for not supporting it and never seem to find the time or resources and thus F# support never happens.

Xamarin.Forms already has support for netstandard projects written in any .net language and there is no plan to remove it. I am sure there are some teams out there that have failed to support F# but how is that relevant to Xamarin.Forms or MAUI?

MAUI has the possibility of improving the experience for developing UI using code, and so having previews/live updates of code-defined UI would benefit F# usage. So it's important that's not done in a C#-specific way. Assuming that happens, the experience will be even better than it is at the moment.

Templates will probably be updated but that's a negligible issue as in Xamarin.Forms/MAUI very little code lives in the platform-specific project(s).

Overall there is a lot of emotion in this thread and it's hard to understand what is causing it.

**Comment by PureWeen** [comment]
We're really excited to work with @TimLariviere so that Fabulous can be developed into a first class citizen with .NET MAUI

https://github.com/fsprojects/Fabulous/issues/747

The work we're doing on the slim renderers
https://github.com/dotnet/maui/issues/28

Is going to allow Fabulous to really structure itself how ever it wants without having to just use wrappers for Forms elements

@TimLariviere has already put together a very descriptive spec on what he's thinking the direction should be

https://github.com/fsprojects/Fabulous/issues/738

If you're interested in F# as a first class citizen for coding UI's I would head over to Fabulous and tell Tim what you think

**Comment by maxenko** [comment]
Chick and the egg thing as well, lack of basic templates keeps a lot of people from even thinking about trying it.

But those of us who use it do know - once you complete a simple project its hard to go back to C#.

**Comment by davidortinau** [comment]
To those who are using F# today with Xamarin.Forms, would you please shoot me a description and perhaps some screenshots of what you're shipping? That'll help me get a clearer picture. david.ortinau@microsoft.com 

Happy to setup some calls as well. :)

**Comment by TimLariviere** [comment]
I would be really happy to help shape MAUI so we can remove all the internal MVVM -> MVU mapping happening inside Fabulous. This would be hugely beneficial, both in terms of performance and time/work necessary for a new release supporting the latest version of MAUI/XF.

That said, not every F# Xamarin developers use Fabulous.

One major pain today, when it comes to using F# for building Xamarin/XF apps, is the F# tooling lagging behind or all the little things missing that would make life so much easier, like updated official templates and samples.

It's not directly linked to MAUI, but it is a mandatory part before F# developers can even think about MAUI.

Examples of this today are:
- [Mono supporting only F# 4.5 (released in 2018), when 4.7 is available](https://github.com/mono/mono/issues/16763)
- [PackageReference not supported in old-style F# projects](https://github.com/dotnet/fsharp/issues/5849)
- No official F# samples (I tried to migrate Fabulous's F# Android template to AndroidX, but was met with a lot of guessing work and frictions)

So, since MAUI will introduce a new project type and I guess lots of tooling around it like Hot Reload and such, it is important to make sure those work with F# like they will do with C#.

Though I have to note that Xamarin is one of the few places where F# is well considered and most likely get support... at least one day.

**Comment by Happypig375** [comment]
This has become the most liked issue in this repository in 2 days! 🎉🚀

@TimLariviere
>I tried to migrate Fabulous's F# Android template to AndroidX, but was met with a lot of guessing work and frictions

The official C# Xamarin.Forms project template doesn't use AndroidX either. Its `TargetFramework` is still at `v9.0`, which indicates the use of Android Support libraries instead of AndroidX. However, I have successfully upgraded my Fabulous project to AndroidX through manual editing of `.fsproj`s, `packages.config`s and Android `.axml`s (and searching StackOverflow), will try opening a pull request today.

**Comment by nbevans** [comment]
It is exceptionally rare to witness a F# thread, in a non-F# repo, get this much activity and upvotes in such a short space of time. I always knew F# and Xamarin had a special relationship of sorts and this proves it without question.

@davidortinau Thank you (and @TimLariviere) for responding on here - it is good to know we have your attention at this primordial moment for MAUI. In terms of a description/screenshots etc I will see what I can put together and get it over to you.

@TimLariviere
> So, since MAUI will introduce a new project type and I guess lots of tooling around it like Hot Reload and such, it is important to make sure those work with F# like they will do with C#.

This was my primary concern. The screenshots on the Build blog post showed a project system dedicated to MAUI which is entirely new. This new project system will allow all platforms to sit under one project - as opposed to one project per platform. It remains unclear to me the extent of how "forced" this change is or isn't. But without further info I had to assume it was a significant breaking change. And we all know the story of Roslyn vs F#'s project system - there is clearly a high risk that this new project system would only be implemented to support Roslyn.

**Comment by Happypig375** [comment]
@nbevans I think that this issue must have circulated around twitter/reddit/slack/... which other issues did not.

**Comment by JamesRandall** [comment]
@davidortinau  I'd be happy to provide some feedback on my usage of F# with Xamarin Forms - my work has been all through Fabulous. Give me a shout if you'd like a conversation about the apps themselves and more detail.

**Comment by vshapenko** [comment]
As an f# dev currently developing on Fabulous, i would be greatly appreciated if MAUI had f# support from day one. 
If details are needed about an app and problems we are experiencing, i could provide it, @davidortinau

**Comment by fwaris** [comment]
Recently I used F#/Fabulous for the very first time and already I know this is how I would like to build all future applications (see https://github.com/fwaris/BertQA).

Most abstractions are leaky. However in Fabulous I was able to construct the app by only relying on the mental model afforded by the abstraction. Did not need to know (still don't know) how it all works under the covers.

**Comment by r4mp4l** [comment]
Please include F#

**Comment by charlesroddie** [comment]
> @TimLariviere So, since MAUI will introduce a new project type and I guess lots of tooling around it like Hot Reload and such, it is important to make sure those work with F# like they will do with C#.

I don't think the new project type as such will make a difference because there will be [backwards compatibility](https://github.com/dotnet/maui/issues/70#issuecomment-632978443) with "existing" projects.

Preview or hot reload would be very valuable for F# and C#-defined UI, and it's important that they do this in a way that is [.net](https://github.com/xamarin/Xamarin.Forms/issues/9166#event-3157403646) rather than [c# - specific](https://developercommunity.visualstudio.com/idea/650684/c-hot-reload-xamarin.html). Anything that requires an interpreter might be tricky.

**Comment by TimLariviere** [comment]
@charlesroddie It's one thing to be backward compatible, it's another to have the new project type fully supported.
F# developers shouldn't be stuck with 3 or more projects because the F# tooling inside Visual Studio and other IDEs didn't follow and they only checked for `csproj` to allow a single project.

An example of this is `PackageReference`. Old-style C# projects allow the use of `PackageReference` in place of `packages.config`, old-style F# projects don't because "it has entered legacy mode, use new style projects now".
Though, as of today, Xamarin projects still use the old-style. And it has already been a long time.

**Comment by charlesroddie** [comment]
SDK-style platform-specific projects [will come with .net 5](https://github.com/dotnet/runtime/issues/34173).

The intent of XF/MAUI is to do things in a platform-independent way most of the time. E.g. we have 98% of our XF .Net code as F# code in netstandard libraries, and the remaining 2% as C# living in platform projects, mostly boilerplate. Converting that 2% to F# doesn't seem like a big deal to me. The XF/MAUI/F# teams have more important things to do IMO.

**Comment by TimLariviere** [comment]
It still leaves the message that C# is the only important thing. If it happens to work for F# too, that's great. If it doesn't, meh.
I understand your point, but people who might be interested by F# will turn away if the tooling is lacking compared to a language they have a high chance of knowing already.

And suddenly, when you only have a single project, you're forced to either support 100% or not at all.

**Comment by Liminiens** [comment]
As a developer on an application written with Fabulous, I really hope that maui will progress with F# in mind from the first days of development.

**Reply by saint4eva** [reply]
Why not stick to fabulous?

**Comment by DualBrain** [comment]
Please let me say how much a appreciate some of the responses here... it truly helps to reinforce what I've come to believe about the F# community!  The VB community can learn so much from the F# community if they would just take the time to do so. 

Not that I have any sway in the matter... but I will share what @Padanian stated was uncalled for and he should really apologize.  His comments are regurgitating the negative behavior that I'm sure he wishes he didn't have to face from C# devs toward VB.  It's uncalled for, it's unnecessary and it's very much counter productive to getting any sort of positive change.

With that out of the way... let me say a public thank you to the F# community for continuing to help remind Microsoft that the CLR doesn't stand for "C# Runtime Library". ;-)  It would be helpful if those building the tooling and stacks on top of .NET would also keep this in mind when designing their stuff.

(And, yes, I'd love to see support for VB (AND F#) in MAUI - if they are going to evolve / redesign; maybe now would be a good time to design it with support for multiple languages from the go.)

To my F# brothers and sisters; keep fighting the good fight!

**Comment by Happypig375** [comment]
![image](https://user-images.githubusercontent.com/19922066/83427418-4c79f500-a463-11ea-9284-8e591939f96b.png)
#Unexpected

**Comment by DualBrain** [comment]
@Happypig375 I would suggest that 100% of the thumb downs on your comment would fall into the  "#Unexpected" category. ;-)

**Comment by Padanian** [comment]
> but I will share what @Padanian stated was uncalled for and he should really apologize. 

I won't. If anything, I will reiterate my opinion.

**Comment by cartermp** [comment]
Just to make some of this concrete:

* Project tooling support for F# _should_ come "for free" because things are based on the the ".NET project system" (also known as "SDK-style project system", or CPS). F# already works there.
* Templates would need to be figured out. They're not expensive to create, but they are a fixed cost to maintain. We've had a pretty good and easy time maintaining F# templates for ASP.NET, so I don't think it would be too expensive doing that here. What will take more time is agreeing on the scope of the templates.
* Visual designer tooling, if it exists at all, will be unlikely to happen for F#. These are incredibly expensive to build and maintain even for C# (and it shouldn't be a surprise that there hasn't been much focus on visual designer tooling in VS for many years...)

**Comment by uxsoft** [comment]
I wouldn't like to speak for all F# devs, but for me personally visual designer tooling for MVU is completely irrelevant. 

I'd be very happy if I can go file a new project and have hot reload working with F# & MAUI.

**Comment by fwaris** [comment]
Don't think a visual designer is much help for dynamic UI's. As @uxsoft  said hot reload is very much desired.

**Comment by mauler2025** [comment]
Ability to render mathematics equations in LaTeX format should be covered in the MAUI core library. :-)

**Reply by houstonhaynes** [reply]
#Expected :-)

**Comment by Happypig375** [comment]
@pstricks-fans There is a reason why this functionality is covered by [libraries](https://github.com/verybadcat/CSharpMath). Xamarin.Forms/MAUI only wrap platform-provided native controls (UIKit or Android.Widgets).

**Comment by MattMS** [comment]
The up-votes on this issue are currently over 5x the next nearest (260 vs 50) https://github.com/dotnet/maui/issues/101 ([src](https://github.com/dotnet/maui/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc))
This is really inspiring and I hope the Microsoft acquisition (and usage) of GitHub includes listening to these community signals.

It is disappointing when people are acting like we are in a zero-sum game, when we have F# people supporting VB usage and encouraging them to advocate for more VB support.
I would think that keeping .NET multi-language is in all of our interests.
Considering this issue is labeled as F# support, it may be better to create a separate issue for VB, to gather battle details there.

At the end, I hope we can see all 3 in that `dotnet new` menu 😄

**Comment by Happypig375** [comment]
>The up-votes on this issue are currently over 5x the next nearest

🤓 Closed issues should also be considered, which beings the factor down to ~2.34x. #43 was just closed 3 days ago.

**Comment by fc1943s** [comment]
Thanks for speaking up guys. One day Microsoft will realize how important F# is for the company's future.

**Reply by vshapenko** [reply]
I do think that would be good for maui developers to keep in mind that not only mvvm exists and there are other scenarios like ones Fabulous suggests. Also i hope in maui there will not be all these overengineered things made like overcomplexity of renderers architecture due to inheritance issues.

**Comment by bmitc** [comment]
I see that this discussion was converted from an issue. Are there some actionable items that we can define to turn this into an issue to design, implement, and track? What are the blockers?

From what I can gather from the discussions and what I've read, the key points are:

1. F# support for the MAUI single project system, or rather, ensuring there's no blockages preventing F# using the project system.

2. F# project templates for all available patterns or architectures.

3. This is maybe not specific to F# but is highly related, and that is the issue of the MVU architecture as discussed in #118, the [Fabulous repo for MAUI support](https://github.com/fsprojects/Fabulous/issues/747) and [some experiments](https://github.com/fsprojects/Fabulous/issues/738), and [here](https://twitter.com/dsymetweets/status/1262797500949331976). Perhaps naively, my feeling is that bringing Fabulous to MAUI to serve as the basis for the MVU architecture seems wonderful. It seems that the ideal situation, again perhaps naively, is that MVVM and MVU (of the Fabulous kind) are available in MAUI to _both_ C# and F# but potentially MVVM geared towards C# and MVU towards F#. Or it could be the case that there's "backends" for both the MVVM and MVU architectures that have some "front-end" interface that are idiomatic to C# and F#. I'm sort of speaking out of ignorance here, as I don't know the feasibility of this, but I do think bringing MVU ala Fabulous to MAUI would be awesome!

(Note that I left off any mention of visual designers. I don't view them as practical to maintain or even very relevant in F#. The dynamic UI supported by Fabulous is fine and doesn't need designers, _especially_ if live reload is implemented.)

I am willing to volunteer to help, if even for something as simple to help maintain the templates. I know F# but am rather new to Xamarin.Forms (learning through Fabulous at the moment). I'm willing to help for whatever it takes to make F# a first class citizen in MAUI. I have a project in mind, and I sort of view MAUI as my last hope for developing a cross-platform desktop application in F# (Fabulous being too focused on mobile since Xamarin Forms itself is).

**Reply by saint4eva** [reply]
C# will be driving the design. And the MVU pattern would be based on Comet MVU - very smart and nice. Then RX UI, Blazor and F# support would be added.

**Reply by bmitc** [reply]
That information seems opposed to [this comment](https://github.com/dotnet/maui/issues/118#issuecomment-635052337) in the linked issue. Do you have any design or roadmap references for this information? Thanks.

**Reply by saint4eva** [reply]
It is not contradicting anything. F# support will definitely be added in form of fabulous or something else. But the Maui MVU would be based on Comet by James Clancey.

**Reply by csharper2010** [reply]
> But the Maui MVU would be based on Comet by James Clancey.

Is that a guess of yours or was that explicitly decided? Is it written down anywhere?

I'm just curious because maintaining the state in some special base class like in Comet does not deliver the full potential of MVU where state is usually modelled as an immutable structure where detecting changes is possible just with doing reference equality checks (at least that's my understanding). The state and the event handling loop don't have any special dependencies in pure MVU, only the view.

**Comment by Lanayx** [comment]
Wanted to add that comet-style MVU is very similar the model of another F# project - [Sutil](https://github.com/davedawkins/Sutil). The reactive Store (aka State in Maui) worked well to me without using Elmish, so it should be just fine to use comet-mvu in F#. The only thing I'm worried is the need for explicit casting, it will make writing markup in F# annoying (just like it did for writing pipelines in ml.net), so if it is the case then some F# bindings will be required.

**Reply by saint4eva** [reply]
Comet is C# MVU for .NET Maui. For F#, it hopefully going to be Fabulous.

**Reply by Lanayx** [reply]
Well, I personally don't like elm architecture, but I like fsharp and I like sutil and comet-mvu style (also known as frp) as well, so I can't agree with fabulous to be the only option for fsharp developer.

**Reply by knocte** [reply]
@Lanayx FWIW I agree with you.

**Reply by saint4eva** [reply]
@Lanayx For F#, it is going to be Fabulous. While for C# MVU, it is going to be Comet.

**Reply by saint4eva** [reply]
> Well, I personally don't like elm architecture, but I like fsharp and I like sutil and comet-mvu style (also known as frp) as well, so I can't agree with fabulous to be the only option for fsharp developer.

It is going to be one of the options. You can develop yours and take advantage of all the goodies in .NET Maui. But as for Comet-MVU style, it is going to be C#. However, if you would like to write yours, or bring a port of Sutil to .NET Maui, that would be great.

**Comment by jakesteele** [comment]
Please support F# :D

**Comment by knocte** [comment]
FYI, if anyone is seeking unofficial F# support for MAUI/Xamarin.Forms, use these templates: https://github.com/fabulousfx/FSharp.Mobile.Templates (even if it lives in a fabulous-related github org, the templates are not using Fabulous framework)

**Comment by davidbuckleyni** [comment]
This will not happen for day one its a pipe dream maybe second year or so but the team bot even mentioned f# beside blazor nore xf supports it

**Comment by bmitc** [comment]
Can we leave declarations of what is or is not on the development horizon to the official MAUI team? I doubt we hear from them given this was an issue converted to a discussion, but it doesn’t help that non-MAUI team members and seemingly non-F# developers are making hard declarations about what will or will not or should or should not happen.

F# is a first class .NET language. It would be silly to see F# not supported at all ala UWP.

**Comment by ShalokShalom** [comment]
Overwhelming support by the community did seemingly result in zero recognition.

F# is officially seen as a first class citizen within Microsoft, while this translates often into subjective ignorance. 

First, adopting a model from the functional world, then ignoring the functional language. 

Genius 👍🏼

**Reply by cartermp** [reply]
The ~Xamarin~ MAUI team is far too underwater to be able to add support for F# when they know that F# developers who use MAUI will just use a thin C# layer talking to an F# library that does all the work.

**Reply by 3xau1o** [reply]
Sounds like the perfect plan as MAUI seems to continue to support C# only, and to be honest that's fine, XAML and Elmish approaches are quite different, another option would be making a F# template based on something like [MAUI community toolkit markup](https://github.com/CommunityToolkit/Maui.Markup) (NO XAML).

**Reply by TimLariviere** [reply]
Several F# templates for MAUI are already available depending on your preferences. :)

For XAML/MVVM: https://github.com/fsharp-mobile/FSharp.Mobile.Templates
For Elmish/MVU: https://github.com/fsprojects/Fabulous (available on NuGet under the name [Fabulous.MauiControls](https://www.nuget.org/packages?q=Fabulous.MauiControls))

**Reply by 3xau1o** [reply]
@TimLariviere thanks, would be nice something more low level, without Elmish, like an imperative setup for a windows with canvas drawing, custom graphics, game loop...

**Reply by TimLariviere** [reply]
@mindlid For this, I would recommend you start from the XAML F# template and keep only the F# code. This will let you do imperative stuff without depending on an external library.

**Reply by Xyncgas** [reply]
You gotta understand, caring what you are saying reading the developer's feedback, it's too stressful for mental health!

**Comment by yyliveyy** [comment]
When resources are limited, we should concentrate our resources to do one thing.  Instead of spreading resources to do a few things well, you get nothing done.

**Comment by ParkerRedford** [comment]
I am curious how F# Blazor works for this. I can see a lot of benefits from this because F# can make javascript and html type safe.

**Reply by saint4eva** [reply]
Maybe you can fabulous

**Reply by ParkerRedford** [reply]
Fabulous isn't blazor.

I tried bolero, but I don't how to wrap the project into a maui app.

**Reply by saint4eva** [reply]
Maybe asking at Bolero GitHub page, they could give you more useful pieces of information than here?

**Comment by Xyncgas** [comment]
Oh, so a bunch of people is waiting for microsoft to provide F# support for MAUI
I'm using MAUI.Blazor, I created my own custom framework that allows me to build application entirely in F#
Behold 杂鱼♥
![image](https://github.com/dotnet/maui/assets/35312807/2628d302-1b66-43dc-838d-4477f0a63dbc)

**Reply by Xyncgas** [reply]
Creativity is an old friend, microsoft is but someone promising you stuffs but things happen with deliveries, let's remember democracy blessed us with anti-trust laws and freedom and the disruptive creativity brings the energy it needs for this industry to be alive so god bless America let's make America great again !

**Reply by LuohuaRain** [reply]
I really want to learn F#, but it seems much harder than C#.

**Reply by saint4eva** [reply]
> Creativity is an old friend, microsoft is but someone promising you stuffs but things happen with deliveries, let's remember democracy blessed us with anti-trust laws and freedom and the disruptive creativity brings the energy it needs for this industry to be alive so god bless America let's make America great again !

Yet, you are here on a Microsoft GitHub repo. lol

**Reply by ParkerRedford** [reply]
> > Creativity is an old friend, microsoft is but someone promising you stuffs but things happen with deliveries, let's remember democracy blessed us with anti-trust laws and freedom and the disruptive creativity brings the energy it needs for this industry to be alive so god bless America let's make America great again !
> 
> Yet, you are here on a Microsoft GitHub repo. lol

At least he can be on here. Not all of us support the con man.

**Reply by Xyncgas** [reply]
> I really want to learn F#, but it seems much harder than C#.

It's another language you still need to program, you'll find it easy after learning it. I could've written this in C# because I am using .NET objects in F# instead of its own types, like many people probably liking F# for the ability to write lesser code, I wanna implement one-liners that let me have advance StackedLayout so instead of vertical of horizontal I wrote quantum computing inspired structures that use waves to represent number and determine where the next element should be I want the ability to do it

In C#, I have to define what's the type when I am implementing a method, it can be a list of Tuple of condition and functor, or something different while I am modifying the code doing the programming, in F# I never needed to tell the compiler what my types are it knows

**Comment by fwaris** [comment]
FYI. I tested Fabulous MAUI. Initial testing shows that it works well (tested only on Windows). No hitches thus far.

To get running use Visual Studio and add 'maui' workload. Then apply the following.

dotnet new install Fabulous.MauiControls.Templates
dotnet new fabulous-mauicontrols -n MyApp

(I took out android target as I don't have that sdk installed yet)
